bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/test/java/com/google/cloud/bigtable/mirroring/hbase1_x/TestMirroringTable.java

```diff
@@ -0,0 +1,1411 @@
+        Arrays.asList(
+            createPut("i", "f", "q", 1, "1"),
+            createPut("a", "f", "q", 2, "2"),
+            createPut("p", "f", "q", "v"),
+            createGet("g"));
+    mirroringTable.batch(operations, results);
+    executorServiceRule.waitForExecutor();
+    verify(primaryTable, times(1)).batch(operations, results);
+    ArgumentCaptor<List<? extends Row>> argumentCaptor = ArgumentCaptor.forClass(List.class);
+    verify(secondaryTable, times(1)).batch(argumentCaptor.capture(), any(Object[].class));
+
+    assertPutsAreEqual(
+        (Put) argumentCaptor.getValue().get(0), (Put) expectedSecondaryOperations.get(0));
+    assertPutsAreEqual(
+        (Put) argumentCaptor.getValue().get(1), (Put) expectedSecondaryOperations.get(1));
+    assertPutsAreEqual(
+        (Put) argumentCaptor.getValue().get(2), (Put) expectedSecondaryOperations.get(2));
+    assertThat(argumentCaptor.getValue().get(3)).isEqualTo(expectedSecondaryOperations.get(3));
+  }
+
+  @Test
+  public void testConcurrentWritesAreFlowControlledBeforePrimaryAction()
+      throws IOException, InterruptedException {
+    this.mirroringTable =
+        spy(
+            new MirroringTable(
+                primaryTable,
+                secondaryTable,
+                this.executorServiceRule.executorService,
+                mismatchDetector,
+                flowController,
+                secondaryWriteErrorConsumer,
+                new ReadSampler(100),
+                true,
+                new MirroringTracer()));
+
+    Put put1 = createPut("r1", "f1", "q1", "v1");
+
+    // Both batches should be called even if first one fails.
+    mockBatch(primaryTable, secondaryTable, put1, new IOException());
+
+    InOrder inOrder = Mockito.inOrder(primaryTable, flowController);
+    try {
+      this.mirroringTable.put(put1);
+      fail("should fail");
+    } catch (IOException ignored) {
+    }
+
+    this.executorServiceRule.waitForExecutor();
+
+    inOrder.verify(flowController).asyncRequestResource(any(RequestResourcesDescription.class));
+    inOrder.verify(primaryTable).batch(eq(Arrays.asList(put1)), any(Object[].class));
+
+    verify(primaryTable).batch(eq(Arrays.asList(put1)), any(Object[].class));
+    verify(secondaryTable).batch(eq(Arrays.asList(put1)), any(Object[].class));
+  }
+
+  @Test
+  public void testNonConcurrentOpsWontBePerformedConcurrently()
+      throws IOException, InterruptedException {
+    setupMirroringTableWithDirectExecutor();
+    Get get = createGet("get1");
+    Increment increment = new Increment("row".getBytes());
+    Append append = new Append("row".getBytes());
+
+    Put put = createPut("test1", "f1", "q1", "v1");
+    Delete delete = createDelete("test2");
+
+    mockBatch(
+        primaryTable,
+        secondaryTable,
+        get,
+        createResult(),
+        increment,
+        createResult("row", "v1"),
+        append,
+        createResult("row", "v2"));
+
```

Comment:
It is not clear here if we are premuting over different test cases, or if we expect certain combinations to be sequenctial and certain combinations to be parallel. 

If we can set the expecatations clerly that we expect this to be sequential and why, then verify that its called sequentilally, it will be more readable code.

Reply:
Added a comment.